worlds/sc2/ClientGui.py

```diff
@@ -109,149 +112,173 @@ def build(self):
-                            racial_colors = {
-                                SC2Race.TERRAN: (0.24, 0.84, 0.68),
-                                SC2Race.ZERG: (1, 0.65, 0.37),
-                                SC2Race.PROTOSS: (0.55, 0.7, 1)
-                            }
-                            if race in racial_colors:
-                                mission_button.background_color = racial_colors[race]
-                            mission_button.tooltip_text = tooltip
-                            mission_button.bind(on_press=self.mission_callback)
-                            self.mission_id_to_button[mission_id] = mission_button
-                            category_panel.add_widget(mission_button)
-
-                        category_panel.add_widget(Label(text=""))
-                        mission_layout.add_widget(category_panel)
-                    campaign_layout.add_widget(mission_layout)
-                    self.campaign_panel.add_widget(campaign_layout)
-                self.campaign_panel.height = multi_campaign_layout_height
-
-        elif self.launching:
+        if self.launching:
             assert self.campaign_panel is not None
             self.refresh_from_launching = False
 
             self.campaign_panel.clear_widgets()
-            self.campaign_panel.add_widget(Label(text="Launching Mission: " +
-                                                        lookup_id_to_mission[self.launching].mission_name))
+            self.campaign_panel.add_widget(Label(
+                text="Launching Mission: " + lookup_id_to_mission[self.launching].mission_name
+            ))
             if self.ctx.ui:
                 self.ctx.ui.clear_tooltip()
+            return
+        
+        needs_redraw = (
+            self.last_checked_locations != self.ctx.checked_locations
+            or not self.refresh_from_launching
+            or self.last_data_out_of_date != self.ctx.data_out_of_date
+            or self.first_check
+        )
+        if not needs_redraw:
+            return
+
+        assert self.campaign_panel is not None
+        self.refresh_from_launching = True
+
+        self.campaign_panel.clear_widgets()
+        if self.ctx.data_out_of_date:
+            self.campaign_panel.add_widget(Label(text="", padding=[0, 5, 0, 5]))
+            warning_label = Label(
+                text="[color=d23300]Map/Mod data is out of date. Run /download_data in the client[/color]",
+                markup=True,
+                padding=[0, 25, 0, 25],
+            )
+            self.campaign_scroll_panel.border_on = True
+            self.campaign_panel.add_widget(warning_label)
+        else:
+            self.campaign_scroll_panel.border_on = False
+        self.last_data_out_of_date = self.ctx.data_out_of_date
+        if not self.ctx.mission_req_table:
+            self.campaign_panel.add_widget(Label(text="Connect to a world to see a mission layout here."))
+            return
```

Comment:
Only things above this point are new / reordered in this file. The rest is just a de-indent. I got rid of two indentation levels by just turning some `if` statements into early returns (first for the `self.launching` case moved up from the bottom, then from the `if self.ctx.mission_req_table` check right here).

I don't know why github can't handle a lot of whitespace changes.